File: server-rs/src/routes/rules.rs
use axum::{
    Json, Router,
    extract::{Path, Query, State},
    http::StatusCode,
    routing::{get, post, put},
};
use serde::Deserialize;
use serde_json::{Value, json};

use crate::AppState;
use crate::errors::ApiError;
use crate::models::RepositoryRule;

pub(crate) fn rules_routes() -> Router<AppState> {
    Router::new()
        .route("/api/rules", get(list_rules).post(create_rule))
        .route("/api/rules/{rule_id}", put(update_rule).delete(delete_rule))
        .route("/api/rules/from-comment/{comment_id}", post(pin_comment_as_rule))
        .route("/api/rules/bulk-replace", post(bulk_replace_rules))
        .route("/api/rules/optimize", post(optimize_rules))
}

#[derive(Debug, Deserialize)]
struct RulesQuery {
    #[serde(rename = "repoId")]
    repo_id: Option<String>,
}

async fn list_rules(
    State(state): State<AppState>,
    Query(query): Query<RulesQuery>,
) -> Result<Json<Value>, ApiError> {
    let rules = state
        .db
        .list_rules(query.repo_id.as_deref())
        .map_err(ApiError::internal)?;
    Ok(Json(json!({ "rules": rules })))
}

#[derive(Debug, Deserialize)]
struct CreateRulePayload {
    #[serde(rename = "repoId")]
    repo_id: Option<String>,
    content: String,
}

async fn create_rule(
    State(state): State<AppState>,
    Json(payload): Json<CreateRulePayload>,
) -> Result<(StatusCode, Json<Value>), ApiError> {
    let content = payload.content.trim();
    if content.is_empty() {
        return Err(ApiError::bad_request("Rule content is required."));
    }
    let rule = state
        .db
        .create_rule(payload.repo_id.as_deref(), content, "manual", None)
        .map_err(ApiError::internal)?;
    Ok((StatusCode::CREATED, Json(json!({ "rule": rule }))))
}

#[derive(Debug, Deserialize)]
struct RulePath {
    rule_id: String,
}

#[derive(Debug, Deserialize)]
struct UpdateRulePayload {
    content: String,
}

async fn update_rule(
    State(state): State<AppState>,
    Path(path): Path<RulePath>,
    Json(payload): Json<UpdateRulePayload>,
) -> Result<Json<Value>, ApiError> {
    let content = payload.content.trim();
    if content.is_empty() {
        return Err(ApiError::bad_request("Rule content is required."));
    }
    state
        .db
        .update_rule(&path.rule_id, content)
        .map_err(ApiError::internal)?;
    Ok(Json(json!({ "updated": true })))
}

async fn delete_rule(
    State(state): State<AppState>,
    Path(path): Path<RulePath>,
) -> Result<Json<Value>, ApiError> {
    state
        .db
        .delete_rule(&path.rule_id)
        .map_err(ApiError::internal)?;
    Ok(Json(json!({ "deleted": true })))
}

#[derive(Debug, Deserialize)]
struct CommentPath {
    comment_id: String,
}

#[derive(Debug, Deserialize)]
struct PinCommentPayload {
    #[serde(rename = "repoId")]
    repo_id: Option<String>,
}

async fn pin_comment_as_rule(
    State(state): State<AppState>,
    Path(path): Path<CommentPath>,
    Json(payload): Json<PinCommentPayload>,
) -> Result<(StatusCode, Json<Value>), ApiError> {
    let comment = state
        .db
        .get_review_comment_by_id(&path.comment_id)
        .map_err(ApiError::internal)?
        .ok_or_else(|| ApiError::not_found("Review comment not found."))?;

    let rule = state
        .db
        .create_rule(
            payload.repo_id.as_deref(),
            &comment.comment,
            "review_comment",
            Some(&path.comment_id),
        )
        .map_err(ApiError::internal)?;

    Ok((StatusCode::CREATED, Json(json!({ "rule": rule }))))
}

#[derive(Debug, Deserialize)]
struct BulkReplacePayload {
    #[serde(rename = "repoId")]
    repo_id: Option<String>,
    contents: Vec<String>,
}

async fn bulk_replace_rules(
    State(state): State<AppState>,
    Json(payload): Json<BulkReplacePayload>,
) -> Result<Json<Value>, ApiError> {
    let rules = state
        .db
        .bulk_replace_rules(payload.repo_id.as_deref(), &payload.contents)
        .map_err(ApiError::internal)?;
    Ok(Json(json!({ "rules": rules })))
}

#[derive(Debug, Deserialize)]
struct OptimizePayload {
    #[serde(rename = "repoId")]
    repo_id: Option<String>,
    #[serde(rename = "profileId")]
    profile_id: String,
    instruction: Option<String>,
    /// "global" = only repo_id IS NULL, "repo" = only repo_id = repoId
    scope: Option<String>,
}

async fn optimize_rules(
    State(state): State<AppState>,
    Json(payload): Json<OptimizePayload>,
) -> Result<Json<Value>, ApiError> {
    let all_rules = state
        .db
        .list_rules(payload.repo_id.as_deref())
        .map_err(ApiError::internal)?;

    let scope = payload.scope.as_deref().unwrap_or("all");
    let rules: Vec<_> = match scope {
        "global" => all_rules.into_iter().filter(|r| r.repo_id.is_none()).collect(),
        "repo" => all_rules.into_iter().filter(|r| r.repo_id.is_some()).collect(),
        _ => all_rules,
    };

    if rules.is_empty() {
        return Err(ApiError::bad_request("No rules to optimize."));
    }

    let profile = state
        .db
        .get_agent_profile_by_id(&payload.profile_id)
        .map_err(ApiError::internal)?
        .ok_or_else(|| ApiError::bad_request("Agent profile not found."))?;

    let agent_command = super::shared::build_agent_command(&profile);

    let rules_text = rules
        .iter()
        .enumerate()
        .map(|(i, r)| format!("{}. {}", i + 1, r.content))
        .collect::<Vec<_>>()
        .join("\n");

    let user_instruction = payload
        .instruction
        .as_deref()
        .map(|s| s.trim())
        .filter(|s| !s.is_empty())
        .map(|s| format!("\n\nAdditional user instruction:\n{s}"))
        .unwrap_or_default();

    let prompt = format!(
        "You are optimizing a list of repository rules for a coding agent.\n\n\
         Current rules:\n{rules_text}\n\n\
         Instructions:\n\
         - Merge duplicate or overlapping rules\n\
         - Remove contradictory rules (keep the more specific one)\n\
         - Make each rule concise and actionable\n\
         - Preserve the intent of all rules\n\
         - Return ONLY a JSON array of strings, each string being an optimized rule\n\
         - Example: [\"Always use snake_case for function names\", \"Never modify the auth module directly\"]\n\
         - No markdown fences, no extra text. Just the JSON array.{user_instruction}"
    );

    let result = {
        let cmd = agent_command.clone();
        let p = prompt.clone();
        // Use the first repo path or cwd
        let working_dir = if let Some(ref rid) = payload.repo_id {
            state.db.get_repo_by_id(rid)
                .ok()
                .flatten()
                .map(|r| r.path)
                .unwrap_or_else(|| ".".to_string())
        } else {
            ".".to_string()
        };
        tokio::task::spawn_blocking(move || {
            crate::planner::invoke_agent_cli(&cmd, &p, &working_dir, None, None)
        })
        .await
        .map_err(|e| ApiError::internal(anyhow::anyhow!("spawn error: {}", e)))?
        .map_err(ApiError::internal)?
    };

    // Parse the optimized rules from agent output
    let text = result.text.trim();
    let optimized: Vec<String> = serde_json::from_str(text)
        .or_else(|_| {
            // Try to extract JSON array from the text
            if let Some(start) = text.find('[') {
                if let Some(end) = text.rfind(']') {
                    return serde_json::from_str(&text[start..=end]);
                }
            }
            Err(serde_json::Error::io(std::io::Error::new(
                std::io::ErrorKind::InvalidData,
                "Could not parse optimized rules as JSON array",
            )))
        })
        .map_err(|e| ApiError::bad_request(&format!("Failed to parse AI response: {e}")))?;

    Ok(Json(json!({ "optimized": optimized })))
}

/// Format rules as a prompt section for agent injection.
pub fn format_rules_prompt_section(rules: &[RepositoryRule]) -> String {
    if rules.is_empty() {
        return String::new();
    }
    let lines: Vec<String> = rules.iter().map(|r| format!("- {}", r.content)).collect();
    format!(
        "\n\nRepository Rules (follow these strictly):\n{}\n",
        lines.join("\n")
    )
}

File: server-ts/src/routes/rules.ts
import { Router, type Request, type Response } from 'express';

import { ApiError } from '../errors.js';
import { invokeAgentCli } from '../planner/index.js';
import type { AppState } from '../state.js';
import { buildAgentCommand } from './shared.js';

export function rulesRoutes(): Router {
  const router = Router();

  // GET /api/rules
  router.get('/api/rules', (req: Request, res: Response) => {
    const state = req.app.locals.state as AppState;
    try {
      const repoId = req.query.repoId as string;
      const rules = state.db.listRules(repoId || undefined);
      return res.json({ rules });
    } catch (e) {
      if (e instanceof ApiError) return e.toResponse(res);
      return ApiError.internal(e).toResponse(res);
    }
  });

  // POST /api/rules
  router.post('/api/rules', (req: Request, res: Response) => {
    const state = req.app.locals.state as AppState;
    try {
      const body = req.body as { repoId?: string; content: string };
      const content = (body.content ?? '').trim();
      if (!content) {
        return ApiError.badRequest('Rule content is required.').toResponse(res);
      }

      const rule = state.db.createRule(body.repoId, content, 'manual');
      return res.status(201).json({ rule });
    } catch (e) {
      if (e instanceof ApiError) return e.toResponse(res);
      return ApiError.internal(e).toResponse(res);
    }
  });

  // PUT /api/rules/:rule_id
  router.put('/api/rules/:rule_id', (req: Request, res: Response) => {
    const state = req.app.locals.state as AppState;
    try {
      const ruleId = req.params.rule_id as string;
      const body = req.body as { content: string };
      const content = (body.content ?? '').trim();
      if (!content) {
        return ApiError.badRequest('Rule content is required.').toResponse(res);
      }

      state.db.updateRule(ruleId, content);
      return res.json({ updated: true });
    } catch (e) {
      if (e instanceof ApiError) return e.toResponse(res);
      return ApiError.internal(e).toResponse(res);
    }
  });

  // DELETE /api/rules/:rule_id
  router.delete('/api/rules/:rule_id', (req: Request, res: Response) => {
    const state = req.app.locals.state as AppState;
    try {
      const ruleId = req.params.rule_id as string;
      state.db.deleteRule(ruleId);
      return res.json({ deleted: true });
    } catch (e) {
      if (e instanceof ApiError) return e.toResponse(res);
      return ApiError.internal(e).toResponse(res);
    }
  });

  // POST /api/rules/from-comment/:comment_id
  router.post('/api/rules/from-comment/:comment_id', (req: Request, res: Response) => {
    const state = req.app.locals.state as AppState;
    try {
      const commentId = req.params.comment_id as string;
      const body = req.body as { repoId?: string };

      const comment = state.db.getReviewCommentById(commentId);
      if (!comment) {
        return ApiError.notFound('Review comment not found.').toResponse(res);
      }

      const rule = state.db.createRule(body.repoId, comment.comment, 'review_comment', commentId);
      return res.status(201).json({ rule });
    } catch (e) {
      if (e instanceof ApiError) return e.toResponse(res);
      return ApiError.internal(e).toResponse(res);
    }
  });

  // POST /api/rules/bulk-replace
  router.post('/api/rules/bulk-replace', (req: Request, res: Response) => {
    const state = req.app.locals.state as AppState;
    try {
      const body = req.body as { repoId?: string; contents: string[] };
      const rules = state.db.bulkReplaceRules(body.repoId, body.contents);
      return res.json({ rules });
    } catch (e) {
      if (e instanceof ApiError) return e.toResponse(res);
      return ApiError.internal(e).toResponse(res);
    }
  });

  // POST /api/rules/optimize
  router.post('/api/rules/optimize', async (req: Request, res: Response) => {
    const state = req.app.locals.state as AppState;
    try {
      const body = req.body as {
        repoId?: string;
        profileId: string;
        instruction?: string;
        scope?: string;
      };

      let allRules = state.db.listRules(body.repoId);

      const scope = body.scope ?? 'all';
      if (scope === 'global') {
        allRules = allRules.filter((r) => r.repo_id === null);
      } else if (scope === 'repo') {
        allRules = allRules.filter((r) => r.repo_id !== null);
      }

      if (allRules.length === 0) {
        return ApiError.badRequest('No rules to optimize.').toResponse(res);
      }

      const profile = state.db.getAgentProfileById(body.profileId);
      if (!profile) {
        return ApiError.badRequest('Agent profile not found.').toResponse(res);
      }

      const agentCommand = buildAgentCommand(profile);

      const rulesText = allRules
        .map((r, i) => `${i + 1}. ${r.content}`)
        .join('\n');

      const userInstruction = body.instruction?.trim()
        ? `\n\nAdditional user instruction:\n${body.instruction.trim()}`
        : '';

      const prompt =
        `You are optimizing a list of repository rules for a coding agent.\n\n` +
        `Current rules:\n${rulesText}\n\n` +
        `Instructions:\n` +
        `- Merge duplicate or overlapping rules\n` +
        `- Remove contradictory rules (keep the more specific one)\n` +
        `- Make each rule concise and actionable\n` +
        `- Preserve the intent of all rules\n` +
        `- Return ONLY a JSON array of strings, each string being an optimized rule\n` +
        `- Example: ["Always use snake_case for function names", "Never modify the auth module directly"]\n` +
        `- No markdown fences, no extra text. Just the JSON array.${userInstruction}`;

      let workingDir = '.';
      if (body.repoId) {
        const repo = state.db.getRepoById(body.repoId);
        if (repo) {
          workingDir = repo.path;
        }
      }

      const result = await invokeAgentCli(agentCommand, prompt, workingDir, null, null);

      const text = result.text.trim();
      let optimized: string[];

      try {
        optimized = JSON.parse(text);
      } catch {
        const start = text.indexOf('[');
        const end = text.lastIndexOf(']');
        if (start !== -1 && end !== -1 && end > start) {
          try {
            optimized = JSON.parse(text.substring(start, end + 1));
          } catch (e2) {
            return ApiError.badRequest(`Failed to parse AI response: ${e2 instanceof Error ? e2.message : String(e2)}`).toResponse(res);
          }
        } else {
          return ApiError.badRequest('Could not parse optimized rules as JSON array').toResponse(res);
        }
      }

      return res.json({ optimized });
    } catch (e) {
      if (e instanceof ApiError) return e.toResponse(res);
      return ApiError.internal(e).toResponse(res);
    }
  });

  return router;
}
